File: src/main/java/moe/gensoukyo/lib/util/FastMath.java
package moe.gensoukyo.lib.util;

/**
 * @author ChloePrime
 */
public class FastMath {
    /**
     * Fast Inverse Square Root
     * @param x independent variable x
     * @return 1 / sqrt(x)
     */
    public static float invSqrt(float x) {
        float x2 = 0.5F * x;
        int i = Float.floatToRawIntBits(x);
        i = 0x5f3759df - (i >> 1);
        x = Float.intBitsToFloat(i);
        return x * (1.5F - x2 * x * x);
    }

    /**
     * Fast Inverse Square Root
     * @param x independent variable x
     * @return 1 / sqrt(x)
     */
    public static double invSqrt(double x) {
        double x2 = 0.5 * x;
        long i = Double.doubleToRawLongBits(x);
        i = 0x5fe6eb50c7b537aaL - (i >> 1);
        x = Double.longBitsToDouble(i);
        return x * (1.5 - x2 * x * x);
    }

    /**
     * Fast Inverse Square Root
     * @param x independent variable x
     * @return sqrt(x)
     */
    public static float sqrt(float x) {
        return 1 / invSqrt(x);
    }

    /**
     * Fast Inverse Square Root
     * @param x independent variable x
     * @return sqrt(x)
     */
    public static double sqrt(double x) {
        return 1 / invSqrt(x);
    }
}

File: src/main/kotlin/moe/gensoukyo/lib/util/FastMath.kt
@file:Suppress("unused")

package moe.gensoukyo.lib.util

import javax.vecmath.Vector3d
import javax.vecmath.Vector3f
import kotlin.math.sqrt

fun invSqrt (x: Float)  = FastMath.invSqrt(x)
fun invSqrt (x: Double) = FastMath.invSqrt(x)
fun fastSqrt(x: Float)  = FastMath.sqrt(x)
fun fastSqrt(x: Double) = FastMath.sqrt(x)

fun Vector3f.fastNormalize() {
    val norm = invSqrt(x * x + y * y + z * z)
    x *= norm
    y *= norm
    z *= norm
}

fun Vector3d.fastNormalize() {
    val norm = invSqrt(x * x + y * y + z * z)
    x *= norm
    y *= norm
    z *= norm
}

val Vector3f.length: Float get() {
    return sqrt(x * x + y * y + z * z)
}

val Vector3d.length: Double get() {
    return sqrt(x * x + y * y + z * z)
}